File: 064_LongestCommonSubsequence.py
"""
Topic: Dynamic Programming
Problem: Longest Common Subsequence
Description: Classic DP table for LCS length.
Time: O(n*m)
Space: O(min(n,m)) optimized
"""

from typing import List


def longestCommonSubsequence(text1: str, text2: str) -> int:
    if len(text1) < len(text2):
        text1, text2 = text2, text1
    m, n = len(text1), len(text2)
    prev = [0] * (n + 1)
    for i in range(1, m+1):
        cur = [0] * (n + 1)
        for j in range(1, n+1):
            if text1[i-1] == text2[j-1]:
                cur[j] = prev[j-1] + 1
            else:
                cur[j] = max(prev[j], cur[j-1])
        prev = cur
    return prev[n]


if __name__ == '__main__':
    print(longestCommonSubsequence("abcde", "ace"))  # 3
    print(longestCommonSubsequence("abc", "abc"))  # 3

File: 064_LongestCommonSubsequence.cpp
// Topic: Dynamic Programming
// Longest Common Subsequence: classic 2D DP O(m*n)
// Time Complexity: O(m*n)
// Space Complexity: O(min(m,n)) optimized

#include <bits/stdc++.h>
using namespace std;

int longestCommonSubsequence(string text1, string text2)
{
    if (text1.size() < text2.size())
        swap(text1, text2);
    int m = text1.size(), n = text2.size();
    vector<int> dp(n + 1, 0);
    for (int i = m - 1; i >= 0; --i)
    {
        vector<int> ndp(n + 1, 0);
        for (int j = n - 1; j >= 0; --j)
        {
            if (text1[i] == text2[j])
                ndp[j] = 1 + dp[j + 1];
            else
                ndp[j] = max(dp[j], ndp[j + 1]);
        }
        dp.swap(ndp);
    }
    return dp[0];
}

int main()
{
    cout << longestCommonSubsequence("abcde", "ace") << "\n";  // 3
    cout << longestCommonSubsequence("abc", "abc") << "\n";    // 3
    return 0;
}
